lib/compress/zstd_compress.c

```diff
@@ -2928,10 +3028,57 @@ static size_t ZSTD_buildSeqStore(ZSTD_CCtx* zc, const void* src, size_t srcSize)
                                        zc->appliedParams.useRowMatchFinder,
                                        src, srcSize);
             assert(ldmSeqStore.pos == ldmSeqStore.size);
-        } else {   /* not long range mode */
+        } else if (zc->appliedParams.useExternalMatchFinder) {
+            assert(
+                zc->externalMatchCtx.seqBufferCapacity >= ZSTD_sequenceBound(srcSize)
+            );
+            assert(zc->externalMatchCtx.mFinder != NULL);
+
+            {   size_t const nbExternalSeqs = (zc->externalMatchCtx.mFinder)(
+                    zc->externalMatchCtx.mState,
+                    zc->externalMatchCtx.seqBuffer,
+                    zc->externalMatchCtx.seqBufferCapacity,
+                    src, srcSize,
+                    NULL, 0,  /* dict and dictSize, currently not supported */
+                    zc->appliedParams.compressionLevel
+                );
+
+                size_t const nbPostProcessedSeqs = ZSTD_postProcessExternalMatchFinderResult(
+                    zc->externalMatchCtx.seqBuffer,
+                    nbExternalSeqs,
+                    zc->externalMatchCtx.seqBufferCapacity,
+                    srcSize == 0  /* emptySrc */
+                );
+
+                if (!ZSTD_isError(nbPostProcessedSeqs)) {
+                    ZSTD_sequencePosition seqPos = {0,0,0};
+                    ZSTD_copySequencesToSeqStoreExplicitBlockDelim(
+                        zc, &seqPos, zc->externalMatchCtx.seqBuffer, nbPostProcessedSeqs, src, srcSize
+                    );
+                    ms->ldmSeqStore = NULL;
+                    lastLLSize = 0;
```

Comment:
why is `lastLLSize` automatically `= 0` here ?
Shouldn't it depend on the last sequence ?

Reply:
No, because we require `sum(matchlen) + sum(litlen) == srcSize`

Reply:
> No, because we require `sum(matchlen) + sum(litlen) == srcSize`

OK, you mean, that's the requirement for _the input of_ `ZSTD_copySequencesToSeqStoreExplicitBlockDelim()`, and this method is already taking care of the last "incomplete" sequence (all literals), which is actually named a Block Delimiter in this case, by copying these last literals into the literals portion of the `seqStore`.

That makes sense.
I think it would also help if these properties were mentioned in the documentation of the function, into `zstd_compress_internal.h`.

It then justifies that you want to opt out of a general mechanism which adds the last literals to the `seqStore`, because it was already done inside `ZSTD_copySequencesToSeqStoreExplicitBlockDelim()`,  _in contrast with_ the "regular" software machfinders.
And btw, you could clearly opt out by returning from the function right away, thus making clear that no further parts of the function are applicable to this code path.

Reply:
Yup, it's a constraint on the input of ZSTD_copySequencesToSeqStoreExplicitBlockDelim() as well as the output of the external matchfinder function.

Reply:
My only concern on the early return is that someone might add more code to the end of the function, and forget to remove the early return...